Automattic/jetpack
# [Experiment] Blocks: Add a new shared Button block

Alternative approach to https://github.com/Automattic/jetpack/pull/14920

See #14935

#### Changes proposed in this Pull Request:


In #14920 I've tried creating a Button component, complete with utility functions, ready to be used inside existing blocks, replacing our many re-implementations of the Core's Button block.

Watching some WPBlockTalk talks and talking with @apeatling made me rethink this approach, and try with `InnerBlocks`, which would have many UX improvements over the Button components.

Unfortunately, we still can't use the Core's Button as-is. Our use cases are different, and we need different features, and more flexibility.

Some features of this new `jetpack/button` block:
- Server-side rendering (default).
- Can be saved in post content (optional).
- Can output as `a`, `input type="submit"`, `button type"submit"`. (If saved in post content, it's forced as `a` because otherwise the WPCOM KSES would strip it).
- Can be extended with more attributes, data-attributes, etc.

Before committing to this approach, we should see if it actually works for all blocks currently using buttons.
Examples: #15377

#### Note for @Automattic/explorers 

Most of this PR is copypasted from #14920.
Namely, the Button block files is pretty much `===` the Button component from the other PR, but with adjusted attribute names (`buttonTextColor` becomes `textColor` as it's clear that it refers to the button now).

The big differences are in how it's used by Revue.
This PR is slightly larger because changing from component to inner block makes the deprecation/migration a bit more verbose (especially in the PHP side).

Minor note: there's currently no fill/outline style variations. We can easily add them in a second moment if we want (and we do: Eventbrite uses them!)

#### Testing instructions:






* Create a new post.
* Add a Revue block, and insert a Revue username.
* Make sure the Button inside the Revue block works as expected.
* Save the post and view it.
* Make sure the button looks as expected.
* Try filling the Revue form and submitting it.
* Make sure the form works as expected.
* Deactivate Jetpack and reload the post on the front end.
* Make sure the form has been replaced with a simple link to the Revue profile.

#### Proposed changelog entry for your changes:

* N/A
